File: screenshot/Pages/Tutorial/TutorialEmailSlideView.m
//
//  TutorialEmailSlideView.m
//  screenshot
//
//  Created by Corey Werner on 8/9/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

#import "TutorialEmailSlideView.h"
#import "Geometry.h"
#import "TappableTextView.h"
#import "WebViewController.h"
#import "AnalyticsManager.h"
#import "screenshot-Swift.h"

@interface TutorialEmailSlideView () <UITextFieldDelegate, TappableTextViewDelegate>

@property (nonatomic) BOOL readyToSubmit;

@property (nonatomic, strong) UITextField *nameTextField;
@property (nonatomic, strong) UITextField *emailTextField;
@property (nonatomic, strong) MainButton *button;
@property (nonatomic, strong) TappableTextView *textView;

@property (nonatomic, strong) NSLayoutConstraint *expandableViewHeightConstraint;
@property (nonatomic) CGRect keyboardFrame;

@end

@implementation TutorialEmailSlideView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.titleLabel.text = @"Sign Up";
        self.subtitleLabel.text = @"Fill out your info below";
        
        CGFloat p = [Geometry padding];
        
        UILabel *nameLabel = [[UILabel alloc] init];
        nameLabel.translatesAutoresizingMaskIntoConstraints = NO;
        nameLabel.text = @"Your name:";
        nameLabel.layoutMargins = UIEdgeInsetsMake(0.f, 0.f, -5.f, 0.f);
        [self.contentView addSubview:nameLabel];
        [nameLabel setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
        [nameLabel.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:30.f].active = YES;
        [nameLabel.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.contentView.leadingAnchor].active = YES;
        [nameLabel.trailingAnchor constraintLessThanOrEqualToAnchor:self.contentView.trailingAnchor].active = YES;
        [nameLabel.centerXAnchor constraintEqualToAnchor:self.contentView.centerXAnchor].active = YES;
        NSLayoutConstraint *widthConstraint = [NSLayoutConstraint constraintWithItem:nameLabel attribute:NSLayoutAttributeWidth relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.0f constant:400.f];
        widthConstraint.priority = UILayoutPriorityDefaultHigh;
        widthConstraint.active = YES;
        
        _nameTextField = ({
            UITextField *textField = [[UITextField alloc] init];
            textField.translatesAutoresizingMaskIntoConstraints = NO;
            textField.delegate = self;
            textField.text = [[NSUserDefaults standardUserDefaults] valueForKey:UserDefaultsKeys.name];
            textField.placeholder = @"Enter your name";
            textField.backgroundColor = [UIColor whiteColor];
            textField.borderStyle = UITextBorderStyleRoundedRect;
            textField.returnKeyType = UIReturnKeyNext;
            textField.spellCheckingType = UITextSpellCheckingTypeNo;
            textField.autocorrectionType = UITextAutocorrectionTypeNo;
            textField.layoutMargins = UIEdgeInsetsMake(0.f, 0.f, -p, 0.f);
            [self.contentView addSubview:textField];
            [textField setContentCompressionResistancePriority:UILayoutPriorityDefaultLow forAxis:UILayoutConstraintAxisVertical];
            [textField.topAnchor constraintEqualToAnchor:nameLabel.layoutMarginsGuide.bottomAnchor].active = YES;
            [textField.leadingAnchor constraintEqualToAnchor:nameLabel.leadingAnchor].active = YES;
            [textField.trailingAnchor constraintEqualToAnchor:nameLabel.trailingAnchor].active = YES;
            [textField.heightAnchor constraintEqualToConstant:50.f].active = YES;
            textField;
        });
        
        UILabel *emailLabel = [[UILabel alloc] init];
        emailLabel.translatesAutoresizingMaskIntoConstraints = NO;
        emailLabel.text = @"Your email:";
        emailLabel.layoutMargins = UIEdgeInsetsMake(0.f, 0.f, -5.f, 0.f);
        [self.contentView addSubview:emailLabel];
        [emailLabel setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
        [emailLabel.topAnchor constraintEqualToAnchor:self.nameTextField.layoutMarginsGuide.bottomAnchor constant:20.f].active = YES;
        [emailLabel.leadingAnchor constraintEqualToAnchor:self.nameTextField.leadingAnchor].active = YES;
        [emailLabel.trailingAnchor constraintEqualToAnchor:self.nameTextField.trailingAnchor].active = YES;
        
        _emailTextField = ({
            UITextField *textField = [[UITextField alloc] init];
            textField.translatesAutoresizingMaskIntoConstraints = NO;
            textField.delegate = self;
            textField.text = [[NSUserDefaults standardUserDefaults] valueForKey:UserDefaultsKeys.email];
            textField.placeholder = @"yourname@website.com";
            textField.keyboardType = UIKeyboardTypeEmailAddress;
            textField.backgroundColor = [UIColor whiteColor];
            textField.borderStyle = UITextBorderStyleRoundedRect;
            textField.spellCheckingType = UITextSpellCheckingTypeNo;
            textField.autocorrectionType = UITextAutocorrectionTypeNo;
            textField.autocapitalizationType = UITextAutocapitalizationTypeNone;
            [self.contentView addSubview:textField];
            [textField setContentCompressionResistancePriority:UILayoutPriorityDefaultLow forAxis:UILayoutConstraintAxisVertical];
            [textField.topAnchor constraintEqualToAnchor:emailLabel.layoutMarginsGuide.bottomAnchor].active = YES;
            [textField.leadingAnchor constraintEqualToAnchor:emailLabel.leadingAnchor].active = YES;
            [textField.trailingAnchor constraintEqualToAnchor:emailLabel.trailingAnchor].active = YES;
            [textField.heightAnchor constraintEqualToAnchor:self.nameTextField.heightAnchor].active = YES;
            textField;
        });
        
        _button = ({
            MainButton *button = [MainButton buttonWithType:UIButtonTypeCustom];
            button.translatesAutoresizingMaskIntoConstraints = NO;
            [button setTitle:@"Submit" forState:UIControlStateNormal];
            [button addTarget:self action:@selector(submitEmail) forControlEvents:UIControlEventTouchUpInside];
            [self.contentView addSubview:button];
            [button setContentHuggingPriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
            [button setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
            [button.topAnchor constraintEqualToAnchor:self.emailTextField.bottomAnchor constant:40.f].active = YES;
            [button.leadingAnchor constraintGreaterThanOrEqualToAnchor:self.contentView.leadingAnchor].active = YES;
            [button.trailingAnchor constraintLessThanOrEqualToAnchor:self.contentView.trailingAnchor].active = YES;
            [button.centerXAnchor constraintEqualToAnchor:self.contentView.centerXAnchor].active = YES;
            button;
        });
        
        _textView = ({
            TappableTextView *textView = [[TappableTextView alloc] init];
            textView.delegate = self;
            textView.translatesAutoresizingMaskIntoConstraints = NO;
            textView.backgroundColor = [UIColor clearColor];
            textView.textColor = [UIColor gray6];
            textView.textAlignment = NSTextAlignmentCenter;
            textView.font = [UIFont preferredFontForTextStyle:UIFontTextStyleFootnote];
            textView.editable = NO;
            textView.scrollEnabled = NO;
            textView.scrollsToTop = NO;
            [self.contentView addSubview:textView];
            [textView setContentCompressionResistancePriority:UILayoutPriorityRequired forAxis:UILayoutConstraintAxisVertical];
            [textView.topAnchor constraintGreaterThanOrEqualToAnchor:self.button.bottomAnchor constant:p].active = YES;
            [textView.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor].active = YES;
            [textView.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor].active = YES;
            [textView.trailingAnchor constraintEqualToAnchor:self.contentView.trailingAnchor].active = YES;
            
            NSArray *tappableText = @[@{@"By tapping \"Submit\" above, you agree to our\n ": @NO},
                                      @{@"Terms of Service": @YES},
                                      @{@" and ": @NO},
                                      @{@"Privacy Policy": @YES},
                                      @{@" ": @NO}
                                      ];
            
            NSMutableParagraphStyle *paragraph = [[NSMutableParagraphStyle alloc] init];
            paragraph.alignment = textView.textAlignment;
            
            NSDictionary *attributes = @{NSFontAttributeName: textView.font,
                                         NSParagraphStyleAttributeName: paragraph
                                         };
            
            [textView applyTappableText:tappableText withAttributes:attributes];
            textView;
        });
        
        UIView *expandableView = [[UIView alloc] init];
        expandableView.translatesAutoresizingMaskIntoConstraints = NO;
        [self.contentView addSubview:expandableView];
        [expandableView.bottomAnchor constraintEqualToAnchor:self.contentView.bottomAnchor].active = YES;
        _expandableViewHeightConstraint = [NSLayoutConstraint constraintWithItem:expandableView attribute:NSLayoutAttributeHeight relatedBy:NSLayoutRelationEqual toItem:nil attribute:NSLayoutAttributeNotAnAttribute multiplier:1.f constant:0.f];
        self.expandableViewHeightConstraint.active = YES;
        
        [self addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(resignTextField)]];
    }
    return self;
}

- (void)didEnterSlide {
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardWillShow:) name:UIKeyboardWillShowNotification object:nil];
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(keyboardDidHide:) name:UIKeyboardDidHideNotification object:nil];
}

- (void)willLeaveSlide {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
    
    self.textView.delegate = nil;
}


#pragma mark - Submit

- (void)submitEmail {
    if ([self.emailTextField.text isValidEmail]) {
        self.readyToSubmit = YES;
        
        NSString *trimmedName = [self.nameTextField.text trimWhitespace];
        NSString *trimmedEmail = [self.emailTextField.text trimWhitespace];
        
        [[NSUserDefaults standardUserDefaults] setValue:trimmedName forKey:UserDefaultsKeys.name];
        [[NSUserDefaults standardUserDefaults] setValue:trimmedEmail forKey:UserDefaultsKeys.email];
        [[NSUserDefaults standardUserDefaults] synchronize];
        
        [self informDelegateOfSubmittedEmailIfPossible];
        
        [AnalyticsManager track:@"Submitted email" properties:@{@"name": trimmedName, @"email": trimmedEmail}];
        [AnalyticsManager identify:trimmedEmail name:trimmedName];
        
    } else {
        [self.delegate tutorialEmailSlideViewDidFailValidation:self];
    }
    
    [self.emailTextField resignFirstResponder];
}

- (void)informDelegateOfSubmittedEmailIfPossible {
    if (![self.emailTextField isFirstResponder] && self.readyToSubmit) {
        [self resignTextField];
        [self.delegate tutorialEmailSlideViewDidComplete:self];
    }
}


#pragma mark - Text Field

- (void)resignTextField {
    [self endEditing:YES];
}

- (void)textFieldDidBeginEditing:(UITextField *)textField {
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        // Wait until the keyboardFrame has been set.
        
        CGRect toWindowRect = [self convertRect:self.frame toView:self.window];
        CGFloat bottomOffset = self.window.bounds.size.height - CGRectGetMaxY(toWindowRect) + self.button.bounds.size.height;
        self.expandableViewHeightConstraint.constant = MAX(self.keyboardFrame.size.height - bottomOffset, 0);
        
        [UIView animateWithDuration:0.25 animations:^{
            [self.contentView layoutIfNeeded];
        }];
    });
}

- (void)textFieldDidEndEditing:(UITextField *)textField {
    self.expandableViewHeightConstraint.constant = 0;
    
    [UIView animateWithDuration:0.25 animations:^{
        [self.contentView layoutIfNeeded];
    }];
}

- (BOOL)textFieldShouldReturn:(UITextField *)textField {
    if (textField == self.nameTextField) {
        [self.emailTextField becomeFirstResponder];
        
    } else {
        [textField resignFirstResponder];
    }
    
    return YES;
}


#pragma mark - Text View

- (void)tappableTextView:(TappableTextView *)textView tappedTextAtIndex:(NSUInteger)index {
    if (index == 1) {
        [self.delegate tutorialEmailSlideViewDidTapTermsOfService:self];
        
    } else if (index == 3) {
        [self.delegate tutorialEmailSlideViewDidTapPrivacyPolicy:self];
    }
}


#pragma mark - Keyboard

- (void)keyboardWillShow:(NSNotification *)notification {
    if (self.window) {
        self.keyboardFrame = [[[notification userInfo] valueForKey:UIKeyboardFrameEndUserInfoKey] CGRectValue];
    }
}

- (void)keyboardDidHide:(NSNotification *)notification {
    if (self.window) {
        [self informDelegateOfSubmittedEmailIfPossible];
    }
}


#pragma mark - Alert

- (UIAlertController *)failedAlertController {
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Submission Failed" message:@"Please enter a valid email." preferredStyle:UIAlertControllerStyleAlert];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Ok" style:UIAlertActionStyleDefault handler:nil]];
    return alertController;
}


#pragma mark - Legal

+ (UIViewController *)termsOfServiceViewControllerWithDoneTarget:(id)target doneAction:(SEL)action {
    NSString *title = @"Terms of Service";
    NSURL *url = [NSURL URLWithString:@"http://crazeapp.com/legal/#tos"];
    
    return [self webViewControllerWithTitle:title url:url doneTarget:target doneAction:action];
}

+ (UIViewController *)privacyPolicyViewControllerWithDoneTarget:(id)target doneAction:(SEL)action {
    NSString *title = @"Privacy Policy";
    NSURL *url = [NSURL URLWithString:@"http://crazeapp.com/legal/#privacy"];
    
    return [self webViewControllerWithTitle:title url:url doneTarget:target doneAction:action];
}

+ (UIViewController *)webViewControllerWithTitle:(NSString *)title url:(NSURL *)url doneTarget:(id)target doneAction:(SEL)action {
    WebViewController *viewController = [[WebViewController alloc] init];
    viewController.url = url;
    viewController.toolbarEnabled = NO;
    viewController.navigationItem.title = title;
    viewController.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemDone target:target action:action];
    
    return [[UINavigationController alloc] initWithRootViewController:viewController];
}

@end

File: screenshot/Pages/Tutorial/TutorialEmailSlideView.swift
//
//  TutorialEmailSlideView.swift
//  screenshot
//
//  Created by Jacob Relkin on 10/22/17.
//  Copyright © 2017 crazeapp. All rights reserved.
//

import UIKit

protocol TutorialEmailSlideViewDelegate : class {
    func tutorialEmailSlideViewDidComplete(_ slideView: TutorialEmailSlideView)
    func tutorialEmailSlideViewDidTapTermsOfService(_ slideView: TutorialEmailSlideView)
    func tutorialEmailSlideViewDidTapPrivacyPolicy(_ slideView: TutorialEmailSlideView)
}

public class TutorialEmailSlideView : HelperView {
    weak var delegate: TutorialEmailSlideViewDelegate?
    
    let nameTextField = UITextField()
    let emailTextField = UITextField()
    let textView = TappableTextView()
    let button = MainButton()
    var expandableViewHeightConstraint: NSLayoutConstraint!
    
    var keyboardFrame: CGRect = .zero
    var readyToSubmit: Bool = false
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        
        titleLabel.text = "Sign Up"
        subtitleLabel.text = "Fill out your info belows"
        
        let p = Geometry.padding
        
        let nameLabel = { _ -> UILabel in
            let label = UILabel()
            label.translatesAutoresizingMaskIntoConstraints = false
            label.text = "Your name:"
            label.layoutMargins = UIEdgeInsets(top: 0, left: 0, bottom: -5, right: 0)
            
            contentView.addSubview(label)
            label.setContentHuggingPriority(UILayoutPriorityRequired, for: .vertical)
            
            NSLayoutConstraint.activate([
                label.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 30),
                label.leadingAnchor.constraint(greaterThanOrEqualTo: contentView.leadingAnchor),
                label.trailingAnchor.constraint(lessThanOrEqualTo: contentView.trailingAnchor),
                label.centerXAnchor.constraint(equalTo: contentView.centerXAnchor),
                
                { _ -> NSLayoutConstraint in
                    let width = NSLayoutConstraint(item: label, attribute: .width, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1, constant: 400)
                    width.priority = UILayoutPriorityDefaultHigh
                    return width
                }()
            ])
            
            return label
        }()
        
        ({ _ -> Void in
            let textField = nameTextField
            textField.translatesAutoresizingMaskIntoConstraints = false
            textField.delegate = self
            textField.text = UserDefaults.standard.string(forKey: UserDefaultsKeys.name) ?? ""
            textField.placeholder = "Enter your name"
            textField.backgroundColor = .white
            textField.borderStyle = .roundedRect
            textField.returnKeyType = .next
            textField.spellCheckingType = .no
            textField.autocorrectionType = .no
            textField.layoutMargins = UIEdgeInsets(top: 0, left: 0, bottom: -p, right: 0)
            contentView.addSubview(textField)
            
            textField.setContentHuggingPriority(UILayoutPriorityDefaultLow, for: .vertical)
            
            NSLayoutConstraint.activate([
                textField.topAnchor.constraint(equalTo: nameLabel.layoutMarginsGuide.bottomAnchor),
                textField.leadingAnchor.constraint(equalTo: nameLabel.leadingAnchor),
                textField.trailingAnchor.constraint(equalTo: nameLabel.trailingAnchor),
                textField.heightAnchor.constraint(equalToConstant: 50)
            ])
        })()
        
        let emailLabel = { _ -> UILabel in
            let label = UILabel()
            label.translatesAutoresizingMaskIntoConstraints = false
            label.text = "Your email:"
            label.layoutMargins = UIEdgeInsets(top: 0, left: 0, bottom: -5, right: 0)
            
            contentView.addSubview(label)
            label.setContentHuggingPriority(UILayoutPriorityRequired, for: .vertical)
            
            NSLayoutConstraint.activate([
                label.topAnchor.constraint(equalTo: nameTextField.layoutMarginsGuide.bottomAnchor, constant: 20),
                label.leadingAnchor.constraint(equalTo: nameTextField.leadingAnchor),
                label.trailingAnchor.constraint(equalTo: nameLabel.trailingAnchor)
            ])
            
            return label
        }()
        
        ({ _ in
            emailTextField.translatesAutoresizingMaskIntoConstraints = false
            emailTextField.delegate = self
            emailTextField.text = UserDefaults.standard.string(forKey: UserDefaultsKeys.email) ?? ""
            emailTextField.placeholder = "yourname@website.com"
            emailTextField.keyboardType = .emailAddress
            emailTextField.backgroundColor = .white
            emailTextField.borderStyle = .roundedRect
            emailTextField.spellCheckingType = .no
            emailTextField.autocorrectionType = .no
            emailTextField.autocapitalizationType = .none
            
            contentView.addSubview(emailTextField)
            emailTextField.setContentHuggingPriority(UILayoutPriorityDefaultLow, for: .vertical)
            
            NSLayoutConstraint.activate([
                emailTextField.topAnchor.constraint(equalTo: emailLabel.layoutMarginsGuide.bottomAnchor),
                emailTextField.leadingAnchor.constraint(equalTo: emailLabel.leadingAnchor),
                emailTextField.trailingAnchor.constraint(equalTo: emailLabel.trailingAnchor),
                emailTextField.heightAnchor.constraint(equalTo: nameTextField.heightAnchor)
            ])
        })()

        ({ _ in
            let p2 = Geometry.extendedPadding

            button.translatesAutoresizingMaskIntoConstraints = false
            button.setTitle("Submit", for: .normal)
            button.addTarget(self, action: #selector(submitEmail), for: .touchUpInside)
        
            contentView.addSubview(button)
            button.setContentHuggingPriority(UILayoutPriorityRequired, for: .vertical)
            button.setContentCompressionResistancePriority(UILayoutPriorityRequired, for: .vertical)
            
            NSLayoutConstraint.activate([
                button.topAnchor.constraint(equalTo: emailTextField.bottomAnchor, constant: p2),
                button.leadingAnchor.constraint(greaterThanOrEqualTo: contentView.leadingAnchor),
                button.trailingAnchor.constraint(lessThanOrEqualTo: contentView.trailingAnchor),
                button.centerXAnchor.constraint(equalTo: contentView.centerXAnchor)
            ])
        })()
        
        ({
            textView.delegate = self
            textView.translatesAutoresizingMaskIntoConstraints = false
            textView.backgroundColor = .clear
            textView.textColor = .gray6
            textView.textAlignment = .center
            textView.font = UIFont.preferredFont(forTextStyle: .footnote)
            textView.isEditable = false
            textView.isScrollEnabled = false
            textView.scrollsToTop = false
            
            contentView.addSubview(textView)
            
            textView.setContentCompressionResistancePriority(UILayoutPriorityRequired, for: .vertical)
            
            NSLayoutConstraint.activate([
                textView.topAnchor.constraint(greaterThanOrEqualTo:button.bottomAnchor, constant: p),
                textView.leadingAnchor.constraint(equalTo: contentView.leadingAnchor),
                textView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor),
                textView.trailingAnchor.constraint(equalTo: contentView.trailingAnchor)
            ])
            
            let tappableText:[[String: NSNumber]] = [
                ["By tapping \"Submit\" above, you agree to our\n": NSNumber(value:false)],
                ["Terms of Service": NSNumber(value:true)],
                [" and ": NSNumber(value:false)],
                ["Privacy Policy": NSNumber(value:true)],
                [" ": NSNumber(value:false)]
            ]
            
            let paragraph = NSMutableParagraphStyle()
            paragraph.alignment = textView.textAlignment
            
            let attributes = [ NSFontAttributeName : textView.font!,
                               NSParagraphStyleAttributeName : paragraph ] as [String : Any]
            textView.applyTappableText(tappableText, withAttributes: attributes)
        })()

        ({ _ in
            let expandableView = UIView()
            expandableView.translatesAutoresizingMaskIntoConstraints = false
            contentView.addSubview(expandableView)
            
            expandableView.bottomAnchor.constraint(equalTo: contentView.bottomAnchor).isActive = true
            expandableViewHeightConstraint = NSLayoutConstraint(item: expandableView, attribute: .height, relatedBy: .equal, toItem: nil, attribute: .notAnAttribute, multiplier: 1, constant: 0)
            expandableViewHeightConstraint.isActive = true
        })()
        
        addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(resignTextField)))
    }
    
    public required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
        
        textView.delegate = nil
    }
    
    // MARK: Actions
    
    @objc private func submitEmail() {
        readyToSubmit = true
        
        let trimmedName = nameTextField.text?.trimmingCharacters(in: .whitespaces) ?? ""
        let trimmedEmail = (emailTextField.text?.isValidEmail() ?? false) ? (emailTextField.text?.trimmingCharacters(in: .whitespaces) ?? "") : ""
        
        UserDefaults.standard.set(trimmedName, forKey: UserDefaultsKeys.name)
        UserDefaults.standard.set(trimmedEmail, forKey: UserDefaultsKeys.email)
        
        let user = identify(trimmedName, email: trimmedEmail)
        _ = identify(trimmedName, email: trimmedEmail, tracker: AnalyticsTrackers.branch)
        
        if trimmedEmail.count > 0 {
            track("Submitted email", properties: ["id" : user.identifier,
                                                  "name" : trimmedName,
                                                  "email": trimmedEmail])
        } else {
            track("Submitted blank email", properties: [ "id" : user.identifier,
                                                         "name" : trimmedName])
        }

        if let ambassadorUsername = UserDefaults.standard.string(forKey: UserDefaultsKeys.ambasssadorUsername) {
            track("Referring Ambassador", properties: ["username": ambassadorUsername])
        }
        
        UserDefaults.standard.set(user.identifier, forKey: UserDefaultsKeys.userID)
        UserDefaults.standard.synchronize()
        
        informDelegateOfSubmittedEmailIfPossible()
        emailTextField.resignFirstResponder()
    }
    
    @objc private func resignTextField() {
        endEditing(true)
    }
    
    private func informDelegateOfSubmittedEmailIfPossible() {
        if !emailTextField.isFirstResponder, readyToSubmit {
            resignTextField()
            delegate?.tutorialEmailSlideViewDidComplete(self)
        }
    }
    
    // MARK: Legal
    
    class func termsOfServiceViewController(withDoneTarget target: Any?, action: Selector?) -> UIViewController? {
        guard let url = URL(string: "http://crazeapp.com/legal/#tos") else {
            return nil
        }
        
        let title = "Terms of Service"
        
        return webViewController(withTitle: title, url: url, doneTarget: target, action: action)
    }

    class func privacyPolicyViewController(withDoneTarget target: Any?, action: Selector?) -> UIViewController? {
        guard let url = URL(string: "http://crazeapp.com/legal/#privacy") else {
            return nil
        }
        
        let title = "Privacy Policy"
        
        return webViewController(withTitle: title, url: url, doneTarget: target, action: action)
    }
    
    private class func webViewController(withTitle title: String, url: URL, doneTarget target: Any?, action: Selector?) -> UIViewController? {
        let webVC = WebViewController()
        webVC.url = url
        webVC.toolbarEnabled = false
        webVC.navigationItem.title = title
        webVC.navigationItem.leftBarButtonItem = UIBarButtonItem(barButtonSystemItem: .done, target: target, action: action)
        
        return UINavigationController(rootViewController: webVC)
    }
    
    // MARK: Alert
    
    func failedAlertController() -> UIAlertController {
        let controller = UIAlertController(title: "Submission Failed", message: "Please enter a valid email.", preferredStyle: .alert)
        controller.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
        return controller
    }
    
    // MARK: Keyboard
    
    @objc func keyboardDidHide(_ note: NSNotification) {
        guard let _ = window else {
            return
        }
        
        informDelegateOfSubmittedEmailIfPossible()
    }
    
    @objc func keyboardWillShow(_ note: NSNotification) {
        guard let _ = window,
        let value = note.userInfo?[UIKeyboardFrameEndUserInfoKey] as? NSValue else {
            return
        }
        
        keyboardFrame = value.cgRectValue
    }
}

extension TutorialEmailSlideView : TappableTextViewDelegate {
    public func tappableTextView(_ textView: TappableTextView!, tappedTextAt index: UInt) {
        switch index {
        case 3:
            delegate?.tutorialEmailSlideViewDidTapPrivacyPolicy(self)
        case 1:
            delegate?.tutorialEmailSlideViewDidTapTermsOfService(self)
            break
        default:
            break
        }
    }
}

extension TutorialEmailSlideView : UITextFieldDelegate {
    public func textFieldDidBeginEditing(_ textField: UITextField) {
        // Wait until keyboard frame has been set!
        
        DispatchQueue.main.async {
            guard let window = self.window else {
                return
            }
            
            let toWindowRect = self.convert(self.frame, to: window)
            let bottomOffset = window.bounds.size.height - toWindowRect.maxY + self.button.bounds.size.height
            
            self.expandableViewHeightConstraint.constant = max(self.keyboardFrame.size.height - bottomOffset, 0)
            UIView.animate(withDuration: 0.25) {
                self.contentView.layoutIfNeeded()
            }
        }
    }
    
    public func textFieldDidEndEditing(_ textField: UITextField) {
        expandableViewHeightConstraint.constant = 0
        
        UIView.animate(withDuration: 0.25) {
            self.contentView.layoutIfNeeded()
        }
    }
    
    public func textFieldShouldReturn(_ textField: UITextField) -> Bool {
        guard textField == nameTextField else {
            textField.resignFirstResponder()
            return true
        }
        
        emailTextField.becomeFirstResponder()
        return true
    }
}

extension TutorialEmailSlideView : TutorialSlideView {
    public func didEnterSlide() {
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardWillShow(_:)), name: .UIKeyboardWillShow, object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(keyboardDidHide(_:)), name: .UIKeyboardDidHide, object: nil)
    }
    
    public func willLeaveSlide() {
        NotificationCenter.default.removeObserver(self)
    }
}
